presto-orc/src/main/java/com/facebook/presto/orc/reader/ListStreamReader.java

```diff
@@ -84,20 +86,18 @@ public ListStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStor
     @Override
     public void setReferencedSubfields(List<SubfieldPath> subfields, int depth)
     {
-        HashSet<Long> referencedSubscripts = new HashSet();
+        ImmutableSet.Builder<Integer> referencedSubscripts = ImmutableSet.builder();
+        ImmutableList.Builder<SubfieldPath> pathsForElement = ImmutableList.builder();
         boolean mayPruneElement = true;
-        ArrayList<SubfieldPath> pathsForElement = new ArrayList();
         for (SubfieldPath subfield : subfields) {
             List<PathElement> pathElements = subfield.getPath();
             PathElement subscript = pathElements.get(depth + 1);
-            if (!subscript.getIsSubscript()) {
-                throw new IllegalArgumentException("List reader needs a PathElement with a subscript");
-            }
-            if (subscript.getSubscript() == PathElement.allSubscripts) {
+            checkArgument(subscript instanceof LongSubscript, "List reader needs a PathElement with a subscript");
+            if (subscript == allSubscripts()) {
```

Comment:
Because it can be serialized/deserialized, is it safe to use reference equality?

Reply:
Even if practically it will never happen (not sure), we could ensure it by using @JsonCreator on a static factory which always returns the static constant, and making the constructor private.

Reply:
@tdcmeehan Good point. It is not safe as of right now, but it will be safe once we use custom deserialization from #12571 (it won't be calling new AllSubscripts(), but use allSubscripts() factory method.) That said, I could change this to .equals() and implement AllSubscripts.equals(). What do you think?